### ISSUE TITLE
Threepenny GUI bindings

### ISSUE THREAD

**Issue by ggreif** [turion/rhine#97]

**Comment by ggreif** [comment]
I have successfully funneled `threepenny-gui` [events](http://hackage.haskell.org/package/threepenny-gui-0.8.2.4/docs/Graphics-UI-Threepenny-Events.html) as `rhine` clocks. How would one package up such a thing? Analogously to `rhine-gloss` I would expect.

**Comment by turion** [comment]
Yes, that would be a good approach! Can you upload your code somewhere so we can have a look?

Basically, an event source in `threepenny` should be a rhine clock. I don't understand `threepenny` well enough to say more, but I'll learn!

**Comment by turion** [comment]
The implementation looks good (except for the small issue you already mentioned which is easy to fix), but conceptually I'm a bit unsure. Especially with this `Semigroup` instance, although it looks right. The whole point of type-level clocks schedules is missed whenever clocks that can tick at conceptually different times (such as different events) have the same type (e.g. the same event data type). But threepenny doesn't have type-level distinction between different events, and I wouldn't know how to add it without shoehorning it in with a newtype every time.

**Comment by turion** [comment]
As for finalisation, Rhine doesn't have a good hook for that yet. But that's an issue that we want to solve in dunai on a fundamental level with type state.

**Comment by ggreif** [comment]
I have implemented a new threepenny (DOM) event "resize", which only applies to a window. So it could get a type-level tag `["window","resize"]`. All things considering you'll need to establish a type-level hierarchy reflecting the nesting of DOM objects. I have a recorded BoBKonf2015 talk about something like this.

**Comment by turion** [comment]
Ah, that sounds good! Is that talk available online somewhere? If it is foreseeable which kinds of events occur in threepenny, we could model all of them with DataKinds.

**Comment by ggreif** [comment]
Slides + Video: http://bobkonf.de/2015/greif.html
